From 8775d160c4074cebb1ab5eb2127c70d51a9491c9 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Wed, 20 Feb 2019 17:14:15 +0800
Subject: [PATCH] USB: storage: add bounce limit to request_queue for LPAE

The upstream commit 21e07dba9fb1 ("scsi: reduce use of block bounce buffers")
remove setting the block layer bounce limit, ask dma-mapping code to handle
any DMA limits by using an iommu or swiotlb. Unfortunately the swiotlb is not
supported in 32-bit ARM kernel, it means that 32-bit ARM SOCs can't handle DMA
limits.

The axm5516 32-bit ARM SOC support LPAE feature which means that it can support
larger than 4G memory. In this case, the block subsystem will pass a larger-than-4G
physical address wihch exceeds the USB ehci host controller DMA capbility (32-bit DMA).
So we need setting the block layer DMA bounce limit in the usb storage layer.

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/usb/storage/scsiglue.c |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/usb/storage/scsiglue.c b/drivers/usb/storage/scsiglue.c
index e227bb5..daffd53 100644
--- a/drivers/usb/storage/scsiglue.c
+++ b/drivers/usb/storage/scsiglue.c
@@ -138,6 +138,9 @@ static int slave_configure(struct scsi_device *sdev)
 	 */
 	if (!us->pusb_dev->bus->controller->dma_mask)
 		blk_queue_bounce_limit(sdev->request_queue, BLK_BOUNCE_HIGH);
+	else
+		blk_queue_bounce_limit(sdev->request_queue,
+			*us->pusb_dev->bus->controller->dma_mask);
 
 	/*
 	 * We can't put these settings in slave_alloc() because that gets
-- 
1.7.9.5

